commons-mongo/jvm/src/main/scala/com/avsystem/commons/mongo/BsonReaderInput.scala

```diff
@@ -1,48 +1,76 @@
 package com.avsystem.commons
 package mongo
 
-import com.avsystem.commons.serialization.{FieldInput, ListInput, ObjectInput}
+import com.avsystem.commons.serialization.{ListInput, ObjectInput}
 import com.google.common.collect.AbstractIterator
+import org.bson._
 import org.bson.types.{Decimal128, ObjectId}
-import org.bson.{BsonReader, BsonType}
 
-class BsonReaderInput(br: BsonReader, override val legacyOptionEncoding: Boolean = false) extends BsonInput {
+import _root_.scala.annotation.tailrec
+
+class BsonReaderInput(br: BsonReader, override val legacyOptionEncoding: Boolean = false)
+  extends BsonInput {
+
   override def readNull(): Boolean =
-    br.getCurrentBsonType == BsonType.NULL && {
+    bsonType == BsonType.NULL && {
       br.readNull()
       true
     }
 
-  override def readString(): String = br.readString()
-  override def readBoolean(): Boolean = br.readBoolean()
-  override def readInt(): Int = br.readInt32()
+  override def readString(): String =
+    expect(BsonType.STRING, br.readString())
```

Comment:
Wouldn't these type expectations be validated by `org.bson.AbstractBsonReader#verifyBSONType` when `BsonReader#readXXX` is called anyway? The current tests don't mind me removing `expect` completely, but I assume you'd rather `handleFailures` everywhere.

Reply:
Kind of yes but sometimes I want to call `expect` with multiple BSON types and this will not be covered by `verifyBSONType` (with the error message that I want). Therefore I use `expect` everywhere for consistency.